### PR TITLE
ESP join windows

### PR DESCRIPTION
**Pull request by RutgerVanBeek**
Closes PM-260

## Description

This PR joins the source windows which were created in the previous PR to one datastream. A csv output file is created. The data sources are combined, such that there is one event per joint_state event. Then for each of the other events the latest event is appended. 

You are kind of joining based on the time of the event, but because the times are not exactly equal a trick is applied, with a dummy_key with is a variable that is 1 in every event and an aggregate window which only streams the last event for every source window. 

The temperatures are first joined based on time, because the times for temperature events of different joints are always the same.  

A visualization of the model: 
image

## Changes

- Made the naming of windows more consistent. 
- Joined the temperature source windows based on time
- Joined all the sources based on a dummy key to create one source

### CODE REVIEW

File: march_data_collector/esp_models/march.xml [hunk 1]
```diff
@@ -21,166 +22,317 @@
...
+                  <fields left="dummy_key" right="dummy_key" />
                 </conditions>
               </join>
               <output>
                 <field-selection name="time" source="l_time" />
+                <field-selection name="dummy_key" source="l_dummy_key" />
                 <field-selection name="actual_positions" source="l_actual_positions" />
                 <field-selection name="actual_velocities" source="l_actual_velocities" />
                 <field-selection name="desired_positions" source="l_desired_positions" />
                 <field-selection name="desired_velocities" source="l_desired_velocities" />
```
Comment (Olavhaasie) [inline march_data_collector/esp_models/march.xml]: The window is called `join_temperature`, however the positions and velocities are also outputted?

File: march_data_collector/esp_models/march.xml [hunk 2]
```diff
@@ -4,13 +4,14 @@
...
           <windows>
-            <window-join name="joinTemperature">
+            <window-join name="join_temperature_all">
               <join type="inner">
                 <conditions>
                   <fields left="time" right="time" />
                 </conditions>
               </join>
               <output>
+                <field-selection name="dummy_key" source="l_dummy_key" />
```
Comment (Olavhaasie) [inline march_data_collector/esp_models/march.xml]: Why did you add all the `dummy_key`s?

File: march_data_collector/esp_models/march.xml (hunk 1 above)
Comment (RutgerVanBeek) [reply march_data_collector/esp_models/march.xml]: I mean that the temperature data is added to the base stream of the joint state data. So for every joint state event the last temperature event is appended. So the temperature is appended to the joint state event. Same for IMC, Com, IMU and so on. The last event for each of them is appended to the main event stream.

File: march_data_collector/esp_models/march.xml (hunk 2 above)
Comment (RutgerVanBeek) [reply march_data_collector/esp_models/march.xml]: You can't join the event based on time, because these are not exactly the same. However I actually want to join the events that have similar time. Therefor the aggregatewindow filters the event stream to only stream the last event. The join should then happen always, but because a key is necessary for a join I made dummy keys for every event. This method is based on a suggestion from a person who knows a lot of ESP, which I will also ask for feedback.

**Comment by Olavhaasie** [comment]
Closed since #460 already contains this
